src/basic_hierarchy/common/HierarchyBuilder.java

```diff
@@ -0,0 +1,395 @@
+
+		List<BasicNode> allGroups = new ArrayList<BasicNode>( artificialGroups );
+		allGroups.addAll( groups );
+
+		return allGroups;
+	}
+
+	/**
+	 * {@link #findNearestAncestor(List, String[], int, int)}
+	 */
+	private static BasicNode findNearestAncestor( List<BasicNode> groups, String[] childBranchIds, int nearestHeight )
+	{
+		return findNearestAncestor( groups, childBranchIds, nearestHeight, -1 );
+	}
+
+	/**
+	 * Attempts to find the nearest existing group that can act as an ancestor to the group specified in argument. IF no such
+	 * group could be found, this method returns null.
+	 * <p>
+	 * This method relies on the groups' IDs being correctly formatted and allowing us to infer the parent-child relations.
+	 * </p>
+	 * 
+	 * @param groups
+	 *            list of groups to search in
+	 * @param childBranchIds
+	 *            ID segments of the group for which we're trying to find an ancestor
+	 * @param nearestHeight
+	 *            number of segments of the best candidate we have available currently (can be negative to mean 'none')
+	 * @param maxIndex
+	 *            max index to search to in the list of groups, for bounding purposes (can be negative to perform an unbounded search)
+	 * @return the nearest group that can act as an ancestor, or null if not found
+	 */
+	private static BasicNode findNearestAncestor( List<BasicNode> groups, String[] childBranchIds, int nearestHeight, int maxIndex )
+	{
+		if ( maxIndex < 0 ) {
+			maxIndex = groups.size();
+		}
+
+		BasicNode result = null;
+
+		for ( int i = 0; i < maxIndex; ++i ) {
+			BasicNode parentGroup = groups.get( i );
+			String[] parentBranchIds = getGroupBranchIds( parentGroup );
+
+			if ( parentBranchIds.length > nearestHeight ) {
+				if ( areGroupsRelated( parentBranchIds, childBranchIds ) ) {
+					result = parentGroup;
+					nearestHeight = parentBranchIds.length;
+				}
+			}
+		}
+
+		return result;
+	}
+
+	/**
+	 * Convenience method to split a group's IDs into segments for easier processing.
+	 */
+	private static String[] getGroupBranchIds( Node g )
+	{
+		String[] result = g.getId().split( Constants.HIERARCHY_BRANCH_SEPARATOR_REGEX );
+		// Ignore the first index ('gen')
+		return Arrays.copyOfRange( result, 1, result.length );
+	}
+
+	/**
+	 * {@link #areGroupsDirectlyRelated(String[], String[])}
+	 */
+	private static boolean areGroupsDirectlyRelated( Node parent, Node child )
+	{
+		return areGroupsDirectlyRelated( getGroupBranchIds( parent ), getGroupBranchIds( child ) );
+	}
+
+	/**
+	 * Checks whether the two groups are directly related (parent-child).
+	 * This method returns false if both ids point to the same group.
+	 * 
+	 * @param parentIds
+	 *            ID segments of the group acting as parent
+	 * @param childIds
+	 *            ID segments of the group as child
+	 * @return whether the two groups are in fact in a direct parent-child relationship.
+	 */
+	private static boolean areGroupsDirectlyRelated( String[] parentIds, String[] childIds )
```

Comment:
I would suggest to change the name of this method to relate that we are comparing strings and checking if they are showing a parent->child relation

I don't like, when two methods have the same names with different parameters, because it is the sign that these two methods are doing different things.